### DISCUSSION TITLE
Interesting effects with "solver/active_edge_threshold"

### DISCUSSION THREAD

**Discussion by Gamepro5** [godot-jolt/godot-jolt#648] [General]
Hi there,

This isn't an issue per say, it's more of a discussion on the side effects I'm observing with active_edge_threshold.

Strangely enough, it seems like setting the threshold to zero behaves better than keeping it at 50 for me, but this doesn't make much sense to me. I've recorded [a video](https://youtu.be/63vM6rBb6E4) where I showcase the interesting artifacts this setting creates.

**Comment by mihe** [comment]
If it works better at 0 degrees in your particular case, then that's great.

Keep in mind though that this setting affects `RigidBody3D` as well, so having it set to 0 degrees will likely make dynamic bodies catch on any triangle edges shared by triangles that form anything but a flat plane, although you might be less likely to see that when you have "Contact Speculative Distance" and "Contact Allowed Penetration" set to 0, but having those two set that low can also introduce other weird artifacts. Those two settings have no effect on kinematic queries like `move_and_collide` though.

"Position Correction" controls the rate of depenetration rate for `RigidBody3D`, and is called "Baumgarte Stabilization" in Jolt. From what I understand its purpose is similar to what "Recovery Amount" is for `move_and_collide`/`move_and_slide`, in that it prevents temporal jitter/jank when you have multiple collisions happening, and in general makes for a more stable simulation. Its default value of 20% might be a tad aggressive, but that's the default value in Jolt so I left it at that. This setting also has no effect on kinematic queries like `move_and_collide`.

The reason you're not seeing any difference in that one particular case is because this "Active Edge Detection" mechanism only deals with triangle edge collisions within a single triangle-based shape, like a `ConcavePolygonShape3D` or `HeightMapShape3D`, and not edge collisions in general. In that particular case you have separate convex shapes that have been manually joined together, which would require some type of cross-shape/cross-body welding for them to be treated as cohesive surface, which Jolt does not have at the moment, although it's on the roadmap from what I understand.

In any case, I have no intention of changing the default value of this setting at the moment. It works well enough for `RigidBody3D` and `move_and_slide` at its current value, although it's certainly not perfect, but that's just how things go sometimes.

**Reply by Gamepro5** [reply]
> In any case, I have no intention of changing the default value of this setting at the moment. It works well enough for RigidBody3D and move_and_slide at its current value, although it's certainly not perfect, but that's just how things go sometimes.

Of course, I didn't mean to imply that you should. It's just with this new setting I'm able to do more testing than I was previously able to.

**Comment by Gamepro5** [comment]
Here's something weird that I found. Active edge threshold can cause false negatives, not just false positives.

If the solver/active_edge_threshold is less than 46, no collision is detected, but a collision is detected when it is equal to or above. This is odd because I don't seem to be colliding on the edge of a shape, but I may be wrong.

[mrp.zip](https://github.com/godot-jolt/godot-jolt/files/12965059/mrp.zip)

**Reply by Gamepro5** [reply]
@mihe do you know if this expected behavior?

**Reply by mihe** [reply]
This has to do with the penetration depth not being very precise/accurate when you use something like a box or cylinder without shape margins, which was brought up a couple of times in jrouwe/JoltPhysics#671. This results in the recovery phase of `move_and_collide` being over-exaggerated and thus the collision phase hits nothing.

The workaround is to either enable shape margins or keep increasing the "Recovery Iterations" setting until you get the desired result, at the cost of some performance, as more iterations should be able to compensate for the lack of precision.

**Reply by Gamepro5** [reply]
Thanks for the reply!

From testing I found that increasing recovery iterations doesn't seem to change it.

I wish there was a was a way to reap the benefits of shape marins without having the negative side effect of having rounded edges of sharp edged shapes.

It became very clear to me that this engine really has trouble with sharp edged shapes. It has been a while since I last used godot physics, but I did not remember it struggling as much with sharp shapes. Is this related to the whole "non deterministic" aspect of jolt? What about this physics engine makes it so rough with box shapes and cylinder shapes. Every game released before 2010 seemed to use box shapes for their character controllers, and so I had always assumed that It was a safe and easy choice for computers to process. The Source engine, Minecraft, and the Quake engine all use box shapes. 

@jrouwe has stated that Jolt struggles with sharp edged shapes "Because the contact point is not a single point but a surface (e.g. rectangle or sphere) they are much more likely to give a jittering contact point or errors in contact normal" (https://github.com/jrouwe/JoltPhysics/issues/671#issuecomment-1704056434). It's really unfortunate, because I was going for an old fashioned game feel as a node to the quake era of movement shooters. Rounded edges, whether it's from a capsule shape or the shape margin property bugs me because these older games didn't have behaviours where you would "slide" down steep edges like stairs or cliffs.

This is a rather extreme example, but it's obvious when games use capsule shapes. In this game, it's clear that when it moves the capsule shape down and detects a floor, it prevents the player from falling.
![image](https://github.com/godot-jolt/godot-jolt/assets/23440295/8de81868-803a-406a-935d-ff382ac2a71f)
![image](https://github.com/godot-jolt/godot-jolt/assets/23440295/38d5b5e8-ff55-416f-b740-4a4fa19af1d7)
![PXL_20231029_010144264 MP](https://github.com/godot-jolt/godot-jolt/assets/23440295/73c78852-6a23-416a-b805-0c02c1edb82c)
![PXL_20231029_010433263 MP](https://github.com/godot-jolt/godot-jolt/assets/23440295/99e0e4b7-13e1-4b46-a079-bc026d5e5ccf)

There is just no ambiguity as far as the player is concerned as to when they are on or off an edge with cylinders and box shapes. I understand that there is an ambiguity as far as the collision detection algorithms are concerned, and I really wish I was smart enough to suggest the solutions other physics engines like quake and source have found for this.

Currently, my biggest headaches with jolt either have to do with rounded edges or the inactive edge problems, and other people also seem to have similar-ish problems (https://github.com/godot-jolt/godot-jolt/issues/587, https://github.com/godot-jolt/godot-jolt/issues/634, https://github.com/godot-jolt/godot-jolt/issues/529)

I don't know what "tweaking epsilons" means, but if these issues just boil down to that, perhaps it may be worth investing in exposing these values in the Godot Jolt settings?

**Reply by mihe** [reply]
> It has been a while since I last used godot physics, but I did not remember it struggling as much with sharp shapes. Is this related to the whole "non deterministic" aspect of jolt?

From what I understand it has to do with the types of collision detection that Jolt uses, which is sometimes referred to as GJK-EPA. This is something you'll find in other physics engines as well, such as Bullet, Havok, Rapier, Unity Physics (DOTS) and Unreal's Chaos Physics and I'm sure many other real-time physics engines as well. All of these use something similar to Godot's shape margins, although I'm not sure that all of them allows for setting it to zero like Jolt does.

Godot Physics on the other hand uses something called SAT (with GJK-EPA as a fallback in some cases), which doesn't seem to need this type of margin. I can't speak to what other trade-offs there might be between these two methods, but seeing as how SAT isn't exactly a new technique I'm assuming that the people working on the above mentioned engines had good reasons for going with GJK-EPA instead.

Godot Physics is certainly not without its own issues either, especially its cylinder shape, but whether those are related to its use of SAT I can't say.

> Every game released before 2010 seemed to use box shapes for their character controllers, and so I had always assumed that It was a safe and easy choice for computers to process. The Source engine, Minecraft, and the Quake engine all use box shapes.

When looking at the source code for Quake III Arena, as well as a version of the Source engine that's floating around out there, this strikes me as a bit of an apples-and-oranges comparison. First of all, the queries used in their character controllers were AABB queries and not arbitrary box queries. Second, these AABB queries were done against BSP geometry, so not something triangle-based, which was popular back in the day but which has since largely been left behind, including by Source 2. Third, the normals reported by these AABB queries were not collision/contact normals, like you'd get in Godot, but instead surface normals.

I have no idea how Minecraft's physics work, but I would imagine they can get away with something very primitive, given that the entire world (and probably the player as well) is all just axis-aligned boxes.

> Currently, my biggest headaches with jolt either have to do with rounded edges or the inactive edge problems

Edge collisions (or "ghost collisions") are also not unique to Jolt, and is something you'll find in most (all?) of the above mentioned physics engines, including Godot Physics. The active edge detection in Jolt is one way to mitigate this, and while it seems to work well enough in the general case it's obviously not a panacea for all types of ghost collisions.

> I don't know what "tweaking epsilons" means, but if these issues just boil down to that, perhaps it may be worth investing in exposing these values in the Godot Jolt settings?

If there was some magic setting/value that fixed all of this then I imagine it would have been brought up by now. At this point it is what it is, and I'd rather not spend any more time running around in circles. The most probable solution is for Godot to expose actual surface normals as part of its contact data and for you to incorporate those into your character controller in some form (or adapt it in some other way) which might then allow you to re-enable shape margins. As for the practical steps involved in that, I can't help you.

**Reply by Gamepro5** [reply]
Once again, thank you so much for your wisdom. I learn a lot from talking to you, and it means a lot to me that you continue to provide an in-depth analysis of my monkey brain questions.

Do you think that once jolt replaced Godot physics, you would have the power to expose surface normals since you would not be limited to using GDExtension? If I recall correctly, jolt supports surface normals but since Godot physics doesn't, you can't override a value that doesn't exist for Godot jolt.

**Reply by mihe** [reply]
> Do you think that once jolt replaced Godot physics, you would have the power to expose surface normals since you would not be limited to using GDExtension? If I recall correctly, jolt supports surface normals but since Godot physics doesn't, you can't override a value that doesn't exist for Godot jolt.

Even if Jolt were to be integrated as a first-party engine module it would still likely live alongside Godot Physics for a good long while, if not indefinitely, so any changes to the physics server interface would presumably need to be implemented in both implementations anyway, which means it would be the same amount of work as just requesting/implementing it for Godot Physics now.

**Reply by jrouwe** [reply]
To add my 2c:

For your problem that characters can stand 'on air', the Jolt character mover has a very simple solution:

![image](https://github.com/godot-jolt/godot-jolt/assets/1621693/acfa4f5e-45e1-4b0c-bf52-8d0f127d900a)

I define a plane (red line) and any contact that is above that plane is considered a 'sliding contact' (regardless of angle between normal and up vector) and will cause you to slide down. Since you cannot get a perfectly up pointing normal between a triangle and a capsule in this scenario, you always know in which direction to slide.

W.r.t. why Jolt uses GJK-EPA: This algorithm handles any kind of convex shape whereas to my knowledge SAT works for polyhedral convex shapes (i.e. shapes without curved surfaces). GJK-EPA requires margins to make it fast. Margins allow you to use GJK (fast) most of the time and avoid using EPA (slow) unless the objects are deeply penetrating. On my (rather long) to do list there is a task to implement SAT to see how it compares with GJK-EPA in terms of speed, but this is most likely weeks of work. Personally, I don't think it is going to fix the normal issues that we're seeing as the issues seem to be mostly triggered by the algorithm treating each triangle individually instead of trying to handle connected triangles. Once I finish my Baldur's Gate 3 addiction, that's the next thing I'm planning to look at (mostly in the context of simulation, so it is uncertain if the move_and_collide function will get any benefit).

**Reply by Gamepro5** [reply]
Hey @jrouwe, thanks for giving your input! Glad you're enjoying Baldur's Gate 3, I was planning to get into it since I like WoW, and it seems similar (both are heavily inspired by DND).

Right now I feel like my best bet is to try to try to find a way to get surface normals exposed to the physics server. That would be useful in way more ways than just my character controller, so I think it's worth adding to godot. I'll see if I can't make a feature request on it or talk to the guy who (from what I remember) exposed the physics server to GDNative in the first place.

The solution you are using for your character controller is certainly interesting. Does it support stair stepping, or do you rely on map makers to make sloped invisible brushes on stairs? I'm also not sure how your controller considers slopes to be valid floors, as the contact point would not be at the tip of the capsule and would therefore be above the red line.

Does the "surface normal" in Jolt suffer from the same issues as collision normals? I'm not sure how it works as Godot Physics didn't have it, so I've never played around with it.

**Reply by jrouwe** [reply]
> Does it support stair stepping

Yes it does.

> I'm also not sure how your controller considers slopes to be valid floors, as the contact point would not be at the tip of the capsule and would therefore be above the red line.

You can configure how high the red line is. I picked a value that works with slopes you need to be able to walk on (the contact point is not that far off the center).

> Does the "surface normal" in Jolt suffer from the same issues as collision normals?

The surface normal is calculated from the triangle vertices, so it will always be the same and doesn't depend on the configuration of the box/capsule.
